### PR TITLE
Reload using magazines (part 1)

### PR DESCRIPTION
**Pull request by mugling**
The long term goal of this series of PRs is to implement reloading using magazines. There are however a lot of prior obstacles, especially the inventory and gunmod code, so a series of smaller refactors are appropriate.

This PR implements `item::ammo_capacity()` and `item::ammo_remaining()` for use with tools, guns and gunmods. They are intended to replace directly querying `item::charges` and provide a suitable hook for later adding magazines.

I'm uncertain how much to commit for each PR. I was considering refactor `item::reload` to use the new functions as part of this PR?

### CODE REVIEW

**Comment by Coolthulhu** [comment]
It would be a good idea if ammo capacity could depend on magazine and not just the gun.
That way it wouldn't be "gun with a double magazine mod" but "gun loaded with a double sized magazine".

**Comment by chaosvolt** [comment]
Ah, this will be interesting to see.

**Comment by mugling** [comment]
> It would be a good idea if ammo capacity could depend on magazine and not just the gun.

Yes, this is the plan. I was going to include standard and high capacity magazines and probably craftable ammunition belts. All guns will eventually use a magazine although some will be integral as suggested by @kevingranade on the forums. Eventually in the dim and distant future we could consider supporting multiple magazines (so weapons can consume more than one ammo type concurrently)

@Coolthulhu any guidance as to whether stop and try merging this PR or continue by refactoring in `item::reload`?

**Comment by Coolthulhu** [comment]
You could try getting this one to work, but try not to depend too much on anything related to charges in the code you'll be changing (except few functions) to avoid the situation where you have tons of extra lines to change.

That spare magazine part looks like it could become a lot more work than expected. I'm not sure how do you want to add it in this PR, since spare magazines don't affect ammo count directly.

The idea of accessing charges/max charges through getter is obviously good. Later it may require more arguments to the getter (say, `ammo_remaining( ammo_type )`), but that's for later.

**Comment by mugling** [comment]
> That spare magazine part looks like it could become a lot more work than expected

Unfortunately this needs handling as a special case. It can be deprecated once magazines are added

**Comment by mugling** [comment]
Finished for this part although criticisms welcomed

**Comment by kevingranade** [comment]
Ideally every testable change can go in its own PR.  Pragmatically, if that means the resulting PRs are tiny and you can't get a feel for where it's going, it becomes a problem.  As a terrible but the best available rule of thumb, look at the # of LoC you're touching, once it exceeds 100 or so consider wrapping it up and making a PR.  This incorporates the assumption that "wrapping it up" will touch more lines, I'm not asking for precisely 100 LoC PRs :)
Even more ideally, these testable changes are tested by a unit test included in the PR, which has become workable to do somewhat recently, poke around in tests/ some time.

**Comment by Rivet-the-Zombie** [comment]
What about giving firearms a json field that would allow them to share magazines, for stuff like STANAG compatible rifles and such?

File: src/item.cpp [hunk 1]
```diff
@@ -3967,163 +4011,114 @@ static void eject_casings( player &p, item *reload_target, itype_id casing_type
...
-        max_load *= 2;
-    }
+        if( target->is_gun() || target->is_gunmod() ) {
+            target->set_curammo( *ammo );
+        }
 
-    if (has_flag("ATOMIC_AMMO")) {
-        max_load *= 100;
-    }
+         if( ammo_type() == "plutonium" ) {
```
Comment (BevapDin) [inline src/item.cpp]: The indentation in this function is sometime a bit strange (3 spaces per level), probably a remnant of the old code. As you're touching the code anyway, you might as well fix this (should be 4 spaces per level).

**Comment by mugling** [comment]
2d1974f resolves the conflicts with #14057, which by the way @OzoneH3 is very nice work

**Comment by mugling** [comment]
> What about giving firearms a json field that would allow them to share magazines, for stuff like STANAG compatible rifles and such?

Sure, see #14149

**Comment by mugling** [comment]
There are further changes in the magazines branch of my repo but as these depend on `ammo_remaining` and `ammo_capacity` so I'll PR those separately once this is merged

**Comment by kevingranade** [comment]
Spare magazine is broken.
Install spare magazine, load it and the gun, fire all the rounds out of the gun, hit r to swap, and it fails with "You can't reload your <gun>."

**Comment by mugling** [comment]
Why is the buildbot unhappy?

**Comment by chaosvolt** [comment]
o.o

Ooh, this'll be fun to see in action.

**Pull request merged**